snappy/security.go

```diff
@@ -20,94 +20,302 @@
+}
+
+// securityPolicyType is a kind of securityPolicy, we currently
+// have "apparmor" and "seccomp"
+type securityPolicyType struct {
+	name          string
+	basePolicyDir string
+}
+
+var securityPolicyTypeAppArmor = securityPolicyType{
+	name:          "apparmor",
+	basePolicyDir: "/usr/share/apparmor/easyprof",
+}
+
+var securityPolicyTypeSeccomp = securityPolicyType{
+	name:          "seccomp",
+	basePolicyDir: "/usr/share/seccomp",
+}
+
+func (sp *securityPolicyType) policyDir() string {
+	return filepath.Join(dirs.GlobalRootDir, sp.basePolicyDir)
+}
+
+func (sp *securityPolicyType) frameworkPolicyDir() string {
+	frameworkPolicyDir := filepath.Join(policy.SecBase, sp.name)
+	return filepath.Join(dirs.GlobalRootDir, frameworkPolicyDir)
+}
+
+// findTemplate returns the security template content from the template name.
+func (sp *securityPolicyType) findTemplate(templateName string) (string, error) {
+	if templateName == "" {
+		templateName = defaultTemplateName
 	}
 
-	if t.Template == "" {
-		t.Template = defaultTemplate
+	subdir := filepath.Join("templates", defaultPolicyVendor(), defaultPolicyVersion())
+	systemTemplateDir := filepath.Join(sp.policyDir(), subdir, templateName)
+	fwTemplateDir := filepath.Join(sp.frameworkPolicyDir(), "templates", templateName)
+
+	// Read system and framwork policy, but always prefer system policy
+	fns := []string{systemTemplateDir, fwTemplateDir}
+	for _, fn := range fns {
+		content, err := ioutil.ReadFile(fn)
+		// it is ok if the file does not exists
+		if err != nil && os.IsNotExist(err) {
+			continue
+		}
+		// but any other error is a failure
+		if err != nil {
+			return "", err
+		}
+
+		return string(content), nil
 	}
 
-	outStr, err := json.MarshalIndent(t, "", "  ")
+	return "", &errPolicyNotFound{"template", sp, templateName}
+}
+
+// helper for findSingleCap that implements readlines().
+func readSingleCapFile(fn string) ([]string, error) {
+	p := []string{}
+
+	r, err := os.Open(fn)
 	if err != nil {
 		return nil, err
 	}
+	defer r.Close()
+
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		p = append(p, s.Text())
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
 
-	return outStr, nil
+	return p, nil
 }
 
-func handleApparmor(buildDir string, m *packageYaml, hookName string, s *SecurityDefinitions) error {
-	hasSecPol := s.SecurityPolicy != nil && s.SecurityPolicy.Apparmor != ""
-	hasSecOvr := s.SecurityOverride != nil && s.SecurityOverride.Apparmor != ""
+// findSingleCap returns the security template content for a single
+// security-cap.
+func (sp *securityPolicyType) findSingleCap(capName, systemPolicyDir, fwPolicyDir string) ([]string, error) {
+	found := false
+	p := []string{}
+
+	policyDirs := []string{systemPolicyDir, fwPolicyDir}
+	for _, dir := range policyDirs {
+		fn := filepath.Join(dir, capName)
+		newCaps, err := readSingleCapFile(fn)
+		// its ok if the file does not exist
+		if err != nil && os.IsNotExist(err) {
```

Comment:
There are a couple of instances of this. FWIW, IsNotExist(err) implies err != nil. In general these functions are built to take nil into account, as it's handy in cases like this.